[6.0][capi-media-vision] add new inference api to detect pose landmarks

tyoungroy opened the pull request:
Signed-off-by: Tae-Young Chung <ty83.chung@samsung.com>

### Change Description ###

Add documents to describe new MediaVision inference api [ACR-1574](https://code.sec.samsung.net/jira/browse/ACR-1574)

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   if (error_code != MEDIA_VISION_ERROR_NONE)
+       dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+   ```
+4. Configure `g_engine_config` with a body pose model data and its mapping file. In the following example, TensorFlow Lite model is used and `data.tflite` and its mapping file `data_mapping.txt` are in `<OwnDataPath>`. Model data is available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or [public pose model](https://githu ...
````
Please change \<a name="background">Background\</a> to \[Background](#background)

tyoungroy replied:
done

mijin85cho reviewed (commented):
Before @nawab-samsung's review, I left some comment about the issues I found. For the grammar issue of the sentences I think @nawab-samsung will check it.

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   if (error_code != MEDIA_VISION_ERROR_NONE)
+       dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+   ```
+4. Configure `g_engine_config` with a body pose model data and its mapping file. In the following example, TensorFlow Lite model is used and `data.tflite` and its mapping file `data_mapping.txt` are in `<OwnDataPath>`. Model data is available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or [public pose model](https://githu ...
````
```suggestion
4. Configure `g_engine_config` with a body pose model data and its mapping file. In the following example, TensorFlow Lite model is used, the `data.tflite` and its mapping file `data_mapping.txt` are stored in your `<OwnDataPath>`. The model data is available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or the [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). 
```
I cannot comprehend the following line from point 4, can you please rephrase it such that I understand it better:

**Suppose model data and its mapping file in Section [Background](#background):**

Once you clear my doubt, I will write a final version of this whole text.

tyoungroy replied:
It means that the files are used as the same ones which are described in Section Background.
I revised it

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+      dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+      ```
+
+8. Set`mocap.bvh` and its mapping file`mocap_mapping.txt` to`g_pose` handle to compare and detect pose. Suppose MoCap file and its mapping file in Section <a name="background">Background</a>:
````
```suggestion
8. Set `mocap.bvh` and its mapping file `mocap_mapping.txt` to `g_pose` handle to compare and detect the pose. 
```
I cannot comprehend the following line from point 8, can you please rephrase it such that I understand it better:

**Suppose the MoCap file and its mapping file in Section [Background](#background):**

Once you clear my doubt, I will write a final version of this whole text.

tyoungroy replied:
It means that the files are used as the same ones which are described in Section Background.
I revised it

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+        dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+    ```
+
+    `mv_inference_pose_landmark_detect()` invokes `_on_pose_landmark_detected_cb()` callback.
````
```suggestion
The `mv_inference_pose_landmark_detect()` invokes the `_on_pose_landmark_detected_cb()` callback.
```

tyoungroy replied:
done

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+    }
+    ```
+
+10. After the detetion is complete, destroy the source, engine configuration, the inference, and the pose handles using `mv_destroy_source()`, `mv_destroy_engine_config()`, `mv_inference_destroy()`, and `mv_pose_destroy()`:
````
```suggestion
10. After the detection is complete, destroy the source, engine configuration, the inference, and the pose handles using `mv_destroy_source()`, `mv_destroy_engine_config()`, `mv_inference_destroy()`, and `mv_pose_destroy()`:

tyoungroy replied:
done

nawab-samsung reviewed (changes requested):
@tyoungroy 
Request you to kindly incorporate our comments, and also we have certain doubts could you please clarify the same, following which I shall review the files again. 
Thanks @mijin85cho for your suggestions

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+**Figure: Definition of human body pose landmarks and body parts**
+
+![Body pose](./media/mediavision_pose_tizen_def.png)
+Pose landmark detection models are available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public github site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). Those models provides different landmarks information such as the number of landmarks and locations. To use them correctly, you h ...
```
One opinion at here. We don't need to write like "fourteen(14)" when it is used in 14 landmarks.
How about just "14 landmarks"?

Reference : https://developers.google.com/style/numbers#numbers-as-numerals

nawab-samsung replied:
Thanks @mijin85cho 
```suggestion
Pose landmark detection models are available in Open Model Zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public GitHub site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). The public pose models provide landmark information, such as the number of landmarks and locations. To use them correctly, you must map the information to landmarks based on the definition. For example, you can use the [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile), which provides 14 landmarks as follows:

tyoungroy replied:
done

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+![Body pose](./media/mediavision_pose_public_model_def.png),
+
+while `-1` means that there is no landmarks, you can create a mapping file. Suppose that the file is `pose_mapping.txt`, then, you can create `pose_mapping.txt` as
```
I'm not sure with efficiency of changing to list.

Also a question. Can we consider this comment as your opinion to my previous one?

nawab-samsung replied:
@mijin85cho yes sure, going with list is just a suggestion. Yes your opinion is also good. 
Whether we go with a list or not, I have slightly checked the sentences for grammar, conciseness and structure.
I have added a leading line to align the content flow.  Your suggestion will be great on this front. 

Thanks again.

mijin85cho replied:
@nawab-samsung OK, thanks for the checking.

tyoungroy replied:
@nawab-samsung  I revised it. But i'm not sure that list is better or not. In my opinion, it looks better to keep it as a paragraph without list format.

nawab-samsung replied:
@tyoungroy  sure, you both have the same opinion, I agree to your opinions. 
Thanks

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+14
+```
+
+`1` denotes that the first landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_HEAD`.  `-1` at the 3rd denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the 4th denotes that the 3rd landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. You can know how it works with following table.
````
I'm not sure with efficiency of changing to list.

tyoungroy replied:
It is better to keep as a paragraph without list. 
the last sentence ``
The following table shows how the public pose model works with different values, definitions, and mappings: `` makes confusion. 
There are no definition of values, what definitions are, mapping is.
In my opinion,
`` The following table shows how the public pose model works.`` is enough.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+---
+
+Motion capture file (MoCap) includes the movements of objects or people. There are various MoCap formats, but well-known BVH (BioVision Hierarchy) file is supported in Media Vision. BVH file has Hierarchy structure to provide landmark information and the structure can be changed. It means that landmarks information is differenct from the definition. To use them correctly, you have to map those in ...
```
Thanks for the modification. However please check that the link is not working.
The link should be in below form.

\[BVH file](media/mediavision_pose_bvh_sample.bvh)

nawab-samsung replied:
@mijin85cho sure thanks

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+Motion capture file (MoCap) includes the movements of objects or people. There are various MoCap formats, but well-known BVH (BioVision Hierarchy) file is supported in Media Vision. BVH file has Hierarchy structure to provide landmark information and the structure can be changed. It means that landmarks information is differenct from the definition. To use them correctly, you have to map those in ...
+
+![Body pose](./media/mediavision_pose_bvh_sample.png)
+, with 15 landmarks. It starts with Hips and ends with RightRoot. Thus, you can create a mapping file, named of `mocap_mapping.txt` as
```
Opinion to the "The following example starts with Hips and ends with RightRoot with 15 landmarks.".

With current suggestion, the paragraphs will be like this.

---

... For example, the following  [BVH file](file://./media/mediavision_pose_bvh_sample.bvh ) describes a squat pose as follows: 

_\[image]_

The following example starts with Hips and ends with RightRoot with 15 landmarks. You can create a mapping file named `mocap_mapping.txt` as follows: 

---

So in this flow, "follow" is repetitive. How about changing "The following example" to "This example" or similar expression?

nawab-samsung replied:
@mijin85cho thanks, I modified as per your suggestion. 

Now the paragraphs will be like this.

---

... For example, the  [BVH file](file://./media/mediavision_pose_bvh_sample.bvh ) describes a squat pose as follows: 

_\[image]_

The example starts with Hips and ends with RightRoot with 15 landmarks. You can create a mapping file named `mocap_mapping.txt` as follows: 

---

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   ```
+2. Decode the image file and fill the `g_source` handle with the decoded raw data.
+
+   In the following example, `sample.jpg` is the image which a person gets a squat pose and  is in the `<OwnDataPath>` folder.
````
How about removing parentheses of (\`sample.jpg\`)?

nawab-samsung replied:
yes sure, edited it, thanks @mijin85cho

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+`1` denotes that the first landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_HEAD`.  `-1` at the 3rd denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the 4th denotes that the 3rd landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. You can know how it works with following table.
+
+**Table: Example of how  [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile) to map the definition**
```
I think that it is better to keep the link.
It is no problem whether the link is personal forked repository or not.
Even though the link is provided above, it is better to provide the link to readers at the table.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+**Figure: Definition of human body pose landmarks and body parts**
+
+![Body pose](./media/mediavision_pose_tizen_def.png)
+Pose landmark detection models are available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public github site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). Those models provides different landmarks information such as the number of landmarks and locations. To use them correctly, you h ...
```
It is no problem to the link. It is the forked repository but not private. It is  also public repository as the link which i forked from.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+---
+
+Motion capture file (MoCap) includes the movements of objects or people. There are various MoCap formats, but well-known BVH (BioVision Hierarchy) file is supported in Media Vision. BVH file has Hierarchy structure to provide landmark information and the structure can be changed. It means that landmarks information is differenct from the definition. To use them correctly, you have to map those in ...
```
MoCap can includes the movements of people. But, "a person" seem to be much proper at here.
I will revise it.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+RightUpLeg,11
+RightLowLeg,12
+RightFoot,13
+```
````
As your comment there is wrong description. 
I fixed `It starts with hips and ends right foot` to `it starts with hips and ends with left foot`.

Also, 
I revised ` BVH file has Hierarchy structure to provide landmark information the structure can be changed.` with `BVH file has Hierarchy structure to provide landmark information with landmarks' name and the structure can be changed`.

So, `Hips, 10` means that the landmark named of `Hips` in the bvh file corresponds to 10th definition `MV_INFERENCE_HUMAN_POSE_PELVIS`. The order in the mapping file is not important because that it can be mapped by the name.

If there is no mapped landmark, you don't need to list it. For example, index 3 is missing. It means that, in this bvh file, the 3rd definition, `MV_INFERENCE_HUMAN_POSE_THORAX` , is not defined and not used.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+![Body pose](./media/mediavision_pose_tizen_def.png)
+Pose landmark detection models are available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public github site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). Those models provides different landmarks information such as the number of landmarks and locations. To use them correctly, you h ...
+
+![Body pose](./media/mediavision_pose_public_model_def.png),
```
I don't understand your comment. How to separate them? 
do i put the images in one place and refer to it in description?
If you suggest the layout, i will refer to it.

nawab-samsung replied:
@tyoungroy  @mijin85cho just wishes to ensure correct spacing between text and images, as in the view, images are immediately at the end of the text. instead, you must give a line break such that image gets inline with the text.

mijin85cho replied:
> you must give a line break such that image gets inline with the text.

@nawab-samsung I sorry but I don't get your point. 

In the original form, the image and the text lines were not separated as like they were in one line.

_text lines_ (_image_) _text lines_

After @tyoungroy's updated, in my point of view, as like other pages now the images and text lines in this page are separated, not merged. That was what I requested to @tyoungroy to apply.

_text lines_:
(_image_)
_text lines_

Would you please explain your concern again?

nawab-samsung replied:
@mijin85cho I just wished to clear what you had suggested thanks. Exactly as per your last comment.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+
+    ```c
+    #define MODEL_DATA "OwnDataPath/data.tflite"
+    #define MODEL_MAPPING_FILE "OwnDataPath/data_mapping.txt"
````
I used <> in sentences to indicate it includes a path. <OwnDataPath> is equal to OwnDataPath.
In image-classification.md, i used the same expression.
If it causes confusion, i will fix them.

mijin85cho replied:
Question. It is said that in this page \<> are removed from \<OwnDataPath>.
However, in line 154 we have a OwnDataPath with \<>.
`..."/<OwnDataPath>/sample.jpg")...`

To unify the usage of \<>, I think it would be better to choose one way through the page.
We have two ways.

 - remove \<> from the `<OwnDataPath>` on line 154.
 - Apply \<> to every `OwnDataPath`s.

Indeed I prefer the second way. Would you please let me know your opinion?

tyoungroy replied:
I applied <> to every `OwnDataPath`  and at line 154, i changed `"/<OwnDataPath>/sample.jpg"` to `"<OwnDataPath>/sample.jpg`

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+
+    ```c
+    #define MOCAP_DATA "OwnDataPath/mocap.bvh"
+    #define MOCAP_MAPPING_FILE  "OwnDataPath/mocap_mapping.txt"
````
https://github.com/Samsung/tizen-docs/pull/1244#discussion_r491663639

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   if (error_code != MEDIA_VISION_ERROR_NONE)
+       dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+   ```
+4. Configure `g_engine_config` with a body pose model data and its mapping file. In the following example, TensorFlow Lite model is used and `data.tflite` and its mapping file `data_mapping.txt` are in `<OwnDataPath>`. Model data is available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or [public pose model](https://githu ...
````
Yes, it is no problem. It is personal, but not private. it inherits the same license form the repository i forked.

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+    error_code = mv_engine_config_set_array_string_attribute(handle,
+    MV_INFERENCE_OUTPUT_NODE_NAMES,
+    outputNodeName, 1);
```
I added the code to create handle.
and for the simplicity, all error check codes are omitted.

mijin85cho replied:
> and for the simplicity, all error check codes are omitted.

How about adding one line description that the error checking codes are omitted for the simplicity?

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+# Pose Detection
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+| 15 | MV_INFERENCE_HUMAN_POSE_LEFT_KNEE | 13 |
+| 16 | MV_INFERENCE_HUMAN_POSE_LEFT_ANKLE | 14 |
+
+---
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+**Figure: Definition of human body pose landmarks and body parts**
+
+![Body pose](./media/mediavision_pose_tizen_def.png)
+Pose landmark detection models are available in open model zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public github site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). Those models provides different landmarks information such as the number of landmarks and locations. To use them correctly, you h ...
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   ```
+3. To detect landmark of pose from the `sample.jpg`,
+
+    <img src="./media/mediavision_pose_sample_sumo.png" width=300>,
````
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+`1` denotes that the first landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_HEAD`.  `-1` at the 3rd denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the 4th denotes that the 3rd landmark of the model corresponds to `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. You can know how it works with following table.
+
+**Table: Example of how  [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile) to map the definition**
```
Keeping link is no problem. 
Except removing the link, i revised it as your suggestion.

tyoungroy commented on docs/application/native/guides/multimedia/media-vision.md:
```diff
@@ -1,27 +1,28 @@
 # Visual Detection and Recognition
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+# Pose Detection
+
+Pose Detection is new feature of Media Vision Inference API since Tizen 6.0. It can provides landmarks detection. In addition, it defines landmarks and parts of a human body and provides to detect a human pose with an motion capture (MoCap) file which a user can create or edit with tools.
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,342 @@
+
+## [Background](#background)
+
+Tizen defines a human body pose landmarks and body parts as shown in below figure.
```
done

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,342 @@
+   free(dataBuffer);
+   dataBuffer = NULL;
+   ```
+3. To detect landmark of pose from the `sample.jpg`,
````
done

tyoungroy commented:
@mijin85cho  and @nawab-samsung ,
I revised twice based on your comments.
Thank you for your comments.

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+   if (error_code != MEDIA_VISION_ERROR_NONE)
+       dlog_print(DLOG_ERROR, LOG_TAG, "error code= %d", error_code);
+   ```
+2. Decode the image file and fill the `g_source` handle with the decoded raw data.
````
```suggestion
2. Decode the image file and fill the `g_source` handle with the decoded raw data:

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+   if (error_code != MEDIA_VISION_ERROR_NONE)
+       dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+   ```
+4. Configure `g_engine_config` with a body pose model data and its mapping file. Suppose that model data `data.tflite` and its mapping file `data_mapping.txt` are used as the same ones described in [Background](#background) and they are in `<OwnDataPath>`:
````
@tyoungroy I need your opinion on the following change, kindly verify for technical correctness too. 
```suggestion
4. Configure `g_engine_config` with a body pose model data and its mapping file. Suppose that model data `data.tflite` and its mapping file `data_mapping.txt` are used as per the description in the [Background](#background) section and the files are stored in `<OwnDataPath>`:

tyoungroy replied:
How about changing as follows: 
```suggestion
4. Configure `g_engine_config` with a body pose model data and its mapping file. Suppose that model data `data.tflite` and its mapping file `data_mapping.txt`, which are described in the [Background](#background) section, are applied and the files are stored in `<OwnDataPath>`:
```

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+      dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+      ```
+
+8. Suppose that MoCap BVH file `mocap.bvh` and its mapping file `mocap_mapping.txt` are used as the same ones described in [Background](#background) and they are in `<OwnDataPath>`. Then, set them to `g_pose` handle to compare and detect the pose:
````
@tyoungroy 
I need your opinion on the following: 
How about slight modification? Please verify for technical correctness too. 
```suggestion 
8. Suppose that MoCap BVH file `mocap.bvh` and its mapping file `mocap_mapping.txt` are used as per the description in the [Background](#background)section, and the files are stored in `<OwnDataPath>`. Then, set the files to `g_pose` handle for comparing and detecting the pose:

Just a question: 
them refers to ?

tyoungroy replied:
How about changing as follows:
```suggestion
8. Suppose that MoCap BVH file `mocap.bvh` and its mapping file `mocap_mapping.txt`, which are described in the [Background](#background) section, are applied and the files are stored in `<OwnDataPath>`. Then, set the files to `g_pose` handle for comparing and detecting the pose:
```

'them' means the files 'mocap.bvh' and ' mocap_mapping.txt'. They are set to 'g_pose' handle and the they are used to compare the pose with the results from mv_inference_pose_landmark_detect()

nawab-samsung reviewed (changes requested):
@tyoungroy Thanks a lot for incorporating the comments! Much appreciated. 

I have had a look at the changes, a few changes are suggested and a few need your valuable opinions too.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+    error_code = mv_inference_prepare(imagedata.g_inference);
+    if (error_code != MEDIA_VISION_ERROR_NONE)
+      dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+      ```
````
Would you please use same indentation with line 261?

Please remove 2 spaces.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+    error_code = mv_pose_create(&imagedata.g_pose);
+    if (error_code != MEDIA_VISION_ERROR_NONE)
+      dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+      ```
````
Would you please use same indentation with line 269?

Please remove 2 spaces.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,343 @@
+| 16 | MV_INFERENCE_HUMAN_POSE_LEFT_ANKLE | 14 |
+
+
+Motion capture file (MoCap) includes the movements of objects or a person. There are various MoCap formats, but a well-known BVH (BioVision Hierarchy) file is supported in Media Vision. BVH file has Hierarchy structure to provide landmark information with landmarks' name, and the structure can be changed. It means that landmarks information is different from the definition. To use them correctly, ...
```
@nawab-samsung 
Question to "BVH (BioVision Hierarchy) ".
I thought SRI-B's style to write an abbreviation was "BioVision Hierarchy (BVH)".
Would you please check?

nawab-samsung replied:
@mijin85cho thanks,I checked many [sources](https://www.google.com/search?client=firefox-b-d&q=BVHfile), and the BVH file is called mostly and commonly by its abbrevation so kept it otherwise. But as you say, I will go with your suggestion. 

```suggestion
Motion capture file (MoCap) includes the movements of objects or a person. There are various MoCap formats, but a well-known BioVision Hierarchy (BVH) file is supported in Media Vision. BVH file has Hierarchy structure to provide landmark information with landmarks' name, and the structure can be changed. It means that landmarks information is different from the definition. To use them correctly, you have to map the information to the landmarks based on the definition. For example, the [BVH file](/media/mediavision_pose_bvh_sample.bvh) describes a squat pose as follows:

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,343 @@
+        dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+    ```
+
+## Related Information
````
```suggestion
## Related information
```

mijin85cho reviewed (commented):
Thanks for the update. Here are my additional queries with the updated version, and my replies . Would you please check these also?

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,343 @@
+
+    <img src="./media/mediavision_pose_sample_sumo.png" width=300>
+
+   In the following example image `sample.jpg`, a person is shown in a squat pose, and the image is in the `<OwnDataPath>` folder.
```
@mijin85cho thanks for letting me know !

tyoungroy commented:
@nawab-samsung 
I updated and left suggestion https://github.com/Samsung/tizen-docs/pull/1244#discussion_r491812273 and https://github.com/Samsung/tizen-docs/pull/1244#discussion_r493990556 .
Please review them.

@mijin85cho 
I applied your comments. Regarding to https://github.com/Samsung/tizen-docs/pull/1244#discussion_r491925285, 
I added the sentence in a code block as comment, but i'm not sure that it looks good or not. If you give opinion, i will move it.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,347 @@
+14
+```
+
+`1` denotes that the first landmark of the model corresponds to the first definition,  `MV_INFERENCE_HUMAN_POSE_HEAD`. `-1` at the 3rd denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the 4th denotes that the 3rd landmark of the model corresponds to the 4th, `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. The following table shows how the public model works:
````
I missed to check this : Please change to ordinal numbers.

3rd ->third
4th -> fourth

mijin85cho replied:
I don't think we need to add "Similarly". In my opinion indeed they are separate notions and not similar items...

nawab-samsung replied:
Sure, thanks edited as per your suggestion

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,347 @@
+
+`1` denotes that the first landmark of the model corresponds to the first definition,  `MV_INFERENCE_HUMAN_POSE_HEAD`. `-1` at the 3rd denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the 4th denotes that the 3rd landmark of the model corresponds to the 4th, `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. The following table shows how the public model works:
+
+**Table: Example of how  [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile) maps to the definition**
```
How about adding s to odefinition?

...maps to the definitions

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,347 @@
+| 16 | MV_INFERENCE_HUMAN_POSE_LEFT_ANKLE | 14 |
+
+
+MoCap file includes the movements of objects or a person. There are various MoCap formats, but a well-known BioVision Hierarchy (BVH) file is supported in Media Vision. BVH file has a hierarchy structure to provide landmark information with landmarks name, and the structure can be changed. It means that landmarks information is different from the definition. To use the BVH file correctly, you hav ...
```
Question to "landmarks name".

According to the context, isn't one of both right?

 - "landmarks' names"
 - "landmark's name"

In my opinion, it seems most case we'd have multiple landmarks. Thus plural should be used to both "landmark" and "name". Would you please check this and let me know your opinion

mijin85cho replied:
As you can see in the image mediavision_pose_tizen_def.png, weren't there multiple landmarks needed to detect a human pose? While reading the article I thought there should be always multiple landmarks to get how the human pose is. With only one landmark I couldn't imagine how to detect human pose, and that's the reason I asked plural. 
@tyoungroy Would you please also check?

nawab-samsung replied:
@mijin85cho Correct, that's is the intention of the text. And the text is saying exactly what you comprehended. 
 
The word **information** fulfils your point. w.r.t to whatever information is in BVH 

Thanks 
Let us wait for @tyoungroy 's opinion.

tyoungroy replied:
Multiple landmarks are required to detect pose. BVH is a format which has a hierarchy structure to provide landmark information of a person or objects. It can include one landmark but it is useless.  It should includes multiple landmarks if it is meaningful as a format to provide landmark information of a person or objects.
```suggestion
The MoCap file includes the movements of objects or a person. There are various MoCap formats, but a well-known BioVision Hierarchy (BVH) file is supported in Media Vision. BVH file has a hierarchy structure to provide landmark information with landmarks' names, and the structure can be changed. It means that landmark information is different from the definition. To use the BVH file correctly, you have to map the information to the landmarks based on the definitions. For example, the [BVH file](./media/mediavision_pose_bvh_sample.bvh) describes a squat pose as follows:
```
I fixed it to 'landmarks names'

mijin85cho reviewed (commented):
Thanks for the update, @tyoungroy 
Below are my comment with the updated files. Would you please check these?

@nawab-samsung Please also check the files and leave your comments.

nawab-samsung reviewed (changes requested):
@tyoungroy Thanks for incorporating most of our comments and clarifying things wherever needed. I have reviewed the file again a few things require your attention. Please do the needful. 

@mijin85cho As always thanks for your suggestions! Please provide your valuable opinion.

nawab-samsung commented:
@tyoungroy request you to kindly respond with your opinion & please icorporate remaining comments. 

Thanks

tyoungroy commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,347 @@
+RightLowLeg,12
+RightFoot,13
+```
+If there is no mapped landmark, you don't need to list it. For example, index 3 is missing. It means that, in this bvh file, the 3rd definition, `MV_INFERENCE_HUMAN_POSE_THORAX`, is not defined and not used.
````
```suggestion
If there is no mapped landmark, you don't need to list it. For example, index 3 is missing. It means that, in this BVH file, the third definition, `MV_INFERENCE_HUMAN_POSE_THORAX`, is not defined and not used.
```
',' and '`' are not required?

tyoungroy commented:
@nawab-samsung and @mijin85cho 
I updated it.

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,346 @@
+
+**Table: Example of how  [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile) maps to the definitions**
+
+| Value | Defintion | pose_mapping.txt |
```
```suggestion
| Value | Definition | pose_mapping.txt |
```

Can we have a better description for the third column ?

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,346 @@
+
+![Body pose](./media/mediavision_pose_bvh_sample.png)
+
+The example starts with hips and ends with left foot with 15 landmrks. You can create a mapping file named `mocap_mapping.txt` as follows:
```
```suggestion
The example starts with hips and ends with the left foot with 15 landmarks. You can create a mapping file named `mocap_mapping.txt` as follows:

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,346 @@
+RightLowLeg,12
+RightFoot,13
+```
+If there is no mapped landmark, you don't need to list it. For example, index 3 is missing. It means that, in this BVH file, the third definition, `MV_INFERENCE_HUMAN_POSE_THORAX` is not defined and not used.
````
```suggestion 
If there is no mapped landmark, you don't need to list it. For example, index three is missing. It means that, in this BVH file, the third definition, `MV_INFERENCE_HUMAN_POSE_THORAX` is not defined and not used.
```
Would it be good to spell out 3 ?

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,346 @@
+   ```
+2. Decode the image file and fill the `g_source` handle with the decoded raw data:
+
+   In the following example image `sample.jpg`, a person is shown in a squat pose, and the image is in the `<OwnDataPath>` folder.
````
```suggestion
   In the following example image `sample.jpg`, a person is shown in a squat pose, and the image is in the `<OwnDataPath>` folder:
```

tyoungroy replied:
```suggestion
Decode the image file and fill the `g_source` handle with the decoded raw data.
   In the following example image `sample.jpg`, a person is shown in a squat pose, and the image is in the `<OwnDataPath>` folder:
```

nawab-samsung commented on docs/application/native/guides/multimedia/pose-detection.md:
````diff
@@ -0,0 +1,346 @@
+      dlog_print(DLOG_ERROR, LOG_TAG, "error code = %d", error_code);
+    ```
+
+8. Suppose that MoCap BVH file `mocap.bvh` and its mapping file `mocap_mapping.txt`, which are described in the [Background](#background) section, are applied and the files are stored in `<OwnDataPath>`. Then, set the files to `g_pose` handle for comparing and detecting the pose:
````
```suggestion
8. Suppose that the MoCap BVH file `mocap.bvh` and its mapping file `mocap_mapping.txt`, that are described in the [Background](#background) section, are applied and the files are stored in `<OwnDataPath>`. Then, set the files to `g_pose` handle for comparing and detecting the pose:

nawab-samsung reviewed (changes requested):
@tyoungroy.
 A few comments that need your attention, kindly incorporate the comments.
Thanks.

nawab-samsung commented:
@tyoungroy Request you to kindly do the needful. 
Thanks

nawab-samsung reviewed (approved):
@tyoungroy 
Thanks for your support and for clearing our doubts. 

@mijin85cho As always thanks for your opinions and for providing your suggestions.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,345 @@
+
+![Body pose](./media/mediavision_pose_public_model_def.png),
+
+In this model, `-1` denotes that there are no landmarks. Using this landmark information, you can create a mapping file. Suppose you create a mapping file with the name `pose_mapping.txt,` then you can populate the `pose_mapping.txt` file as follows:
```
```suggestion
In this model, `-1` denotes that there are no landmarks. Using this landmark information, you can create a mapping file. Suppose you create a mapping file with the name `pose_mapping.txt`, then you can populate the `pose_mapping.txt` file as follows:
```

nawab-samsung replied:
@tyoungroy 
I think you have missed this comment.

mijin85cho commented on docs/application/native/guides/multimedia/pose-detection.md:
```diff
@@ -0,0 +1,345 @@
+
+![Body pose](./media/mediavision_pose_tizen_def.png)
+
+The Pose landmark detection models are available in Open Model Zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public GitHub site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). The public pose models provide landmark information, such as the number of landmarks and locations. To use them correctly, yo ...
```
With [previous comment](https://github.com/Samsung/tizen-docs/pull/1244#discussion_r491663987), I accepted the link to @tyoungroy's personal repository as kind of official repository for the pose model. This comment is just a question. Is there any reason that you didn't link to the original repository that you forked from?

mijin85cho commented on docs/application/toc_all.md:
```diff
@@ -859,6 +859,7 @@
 ##### [Face Detection, Recognition, and Tracking](/application/native/guides/multimedia/face-detection.md)
```
line 858,
@nawab-samsung, didn't you find out that overview for "Visual Detection and Recognition" is missing with [this comment](https://github.com/Samsung/tizen-docs/pull/1244#discussion_r490043969)?
Are you going to skip the fix for it?

\#### Visual Detection and Recognition
\##### \[Overview](/application/native/guides/multimedia/media-vision.md)
\##### \[Barcode Detection and Generation](/application/native/guides/multimedia/image-barcode.md)

nawab-samsung replied:
@mijin85cho  thanks, will verify and let you know about it. 
Thanks

nawab-samsung replied:
I think there is no overview for it, I am facing problem in viewing this PR. May be GitHub is down. Somehow could view this.
![IMG_20201006_145302](https://user-images.githubusercontent.com/43839737/95188067-94ca6a80-07e9-11eb-83c5-8c1931072ad6.jpg)

nawab-samsung replied:
@mijin85cho 

We already had 

![cc12](https://user-images.githubusercontent.com/43839737/95189995-479bc800-07ec-11eb-813b-15e0a71ad293.PNG)


So I kept it as is, and just suggested to added a child to this. 
Pose detection.

mijin85cho replied:
@nawab-samsung I think you confused **Visual Detection and Recognition** and **Multimedia**.
Please check that **Multimedia** has overview, but  **Visual Detection and Recognition** doesn't, in your image.

According to the toc_all.md file, the pose detection is the child page of **Visual Detection and Recognition**, so the link and description should be added to the overview of **Visual Detection and Recognition**, which doesn't listed up in the menu now.

Also I think you can see https://docs1.stg.tizen.org/staging/1244/application/native/guides/multimedia/media-vision/ page, which is result of media-vision.md in this PR update, is dangling and not included in the left menu bar, because media-vision.md is not listed in the toc_all.md.

nawab-samsung replied:
@mijin85cho  thanks, the change is not done in this PR as the screenshot is from live website. But yes we need to fix this issue. 
Thanks again. 

@tyoungroy  could you please add the file `/application/native/guides/multimedia/media-vision.md` to the Toc as this is the overview file. 
Your Toc will look like this after addition. Kindly verify. 


#### Visual Detection and Recognition
##### [Overview](/application/native/guides/multimedia/media-vision.md)
##### [Barcode Detection and Generation](/application/native/guides/multimedia/image-barcode.md)
##### [Face Detection, Recognition, and Tracking](/application/native/guides/multimedia/face-detection.md)
##### [Image Recognition and Tracking](/application/native/guides/multimedia/image-recognition.md)
##### [Image Classification](/application/native/guides/multimedia/image-classification.md)
##### [Pose Detection](/application/native/guides/multimedia/pose-detection.md)



**Current ToC:** 

#### Visual Detection and Recognition
##### [Barcode Detection and Generation](/application/native/guides/multimedia/image-barcode.md)
##### [Face Detection, Recognition, and Tracking](/application/native/guides/multimedia/face-detection.md)
##### [Image Recognition and Tracking](/application/native/guides/multimedia/image-recognition.md)
##### [Image Classification](/application/native/guides/multimedia/image-classification.md)
 ##### [Pose Detection](/application/native/guides/multimedia/pose-detection.md)

mijin85cho reviewed (commented):
Hello. I checked the latest and got below comments. Would you please check them and reply?

nawab-samsung reviewed (changes requested):
@tyoungroy We need to fix the ToC, please incorporate the latest comment. 

Thanks @mijin85cho

tyoungroy commented:
I added media-vision.md to ToC as an overview. Please check the fix.

nawab-samsung commented:
@tyoungroy thanks a lot, will check and let you if something more is needed.

nawab-samsung commented:
![cc124](https://user-images.githubusercontent.com/43839737/95321441-ae86b300-08b8-11eb-9b93-470b962e29a8.PNG)


@tyoungroy  thanks, seems fine now. 
@mijin85cho  thanks for your suggestion.

nawab-samsung reviewed (changes requested):
@tyoungroy Thanks for your support and for incorporating most of the comments, request you to please click resolve for the below comments, or write your opinion and click resolve. 
Thanks, following which, I will approve the PR.

https://github.com/Samsung/tizen-docs/pull/1244/files#r499500940
https://github.com/Samsung/tizen-docs/pull/1244/files#r499484515

tyoungroy commented:
@nawab-samsung  
I chekced
https://github.com/Samsung/tizen-docs/pull/1244/files#r499500940
https://github.com/Samsung/tizen-docs/pull/1244/files#r499484515
and applied all your comments.

nawab-samsung reviewed (approved):
@tyoungroy thanks for cooperation.
@mijin85cho Thanks for your opinion & support

mijin85cho reviewed (approved):
Thanks for the update and review.
